### ISSUE TITLE
Guice dependency injection into listeners and reporters

### ISSUE THREAD

**Issue by kronar** [cbeust/testng#279]
Is there any way to inject dependencies into Reporters and Listeners (instances of ISuiteListener, IInvokedMethodListener2 etc) ????
It would be a nice feature and as far as i see it should not be very expensive

**Comment by avarabyeu** [comment]
+1 any updates?

**Comment by cbeust** [comment]
Not implemented yet but I just took a look at the source and it looks pretty trivial to do. The code that instantiates the listeners is TestRunner:328 and it uses the listener factory to do so. A simple fix for this would be to define a default factory that looks up if a Guice module was declared and if so, use it to instantiate the listener. If not, just do a regular `newInstance`.

**Comment by avarabyeu** [comment]
@cbeust Could you please implement it? I'm very interested to have this change in the next TestNG release. I think, it will be very helpful for all guys who are using integration with Guice

**Comment by juherr** [comment]
@cbeust I'm not sure the fix will be so easy. Currently, a same listener can be instantiate many times. That means args of the listener could be asked many times which may have bad side effects.

**Comment by mark-weissman** [comment]
Injection into listeners would be super cool. We have a WebDriver instance being injected into tests so that we can save on the cost of open/close of Firefox between tests. We'd like a catch-all onTestFailure() listener that very safely captured the page source and took a screen-shot.  We'll probably hack this with ThreadLocal, but I'm reminded DI was invented in part to help us ordinary devs avoid such traps.

**Comment by kool79** [comment]
**_Disclaimer:_** based on real code. Names of classes and some minor behavior  were changed without compiler verification

**MyInjectableListenerFactory** 
````java
public class MyInjectableListenerFactory implements ITestNGListenerFactory, ITestNGListener {

	private final Map<Class<?>, ITestNGListener> listeners = new HashMap<>();

	/**
	 * Creates listener instance. Injects members. Returns null on failure, so it will be created
	 * by default TestNG factory.
	 */
	@Override
	public ITestNGListener createListener(Class<? extends ITestNGListener> listenerClass) {
		if (!listeners.containsKey(listenerClass)) {
			initListener(listenerClass);
		}
		return listeners.get(listenerClass);
	}

	private void initListener(Class<? extends ITestNGListener> listenerClass) {
		try {
			ITestNGListener listener = listenerClass.getConstructor().newInstance();
			MyModuleFactory.getInjector().injectMembers(listener);
			listeners.put(listenerClass, listener);
		} catch (Exception ex) {
			//init with null, so testng will use default factory
			listeners.put(listenerClass, null);
		}
	}
}
````
**MyModuleFactory**
Used to supply module both for test class and for InjectableListenerFactory (see above).
Creates instance of com.my.super.org.MyDefaultModule overridden by some "root" module, which is  defined with system property "test.rootModule"

````java
public class MyModuleFactory implements IModuleFactory {
	private static Module siteModule;

	public static Injector getInjector() { return Guice.createInjector(getMyModule()); }

	@Override
	public Module createModule(ITestContext context, Class<?> testClass) {
		return getMyModule();
	}

	 /**
	 * @return instance of MyDefaultModule overridden by module from sys-property "test.rootModule"
	 */
	synchronized private static Module getMyModule() {
		if (siteModule == null)
			siteModule = Modules.override(	new com.my.super.org.MyDefaultModule()).with(initRootModule());
		return siteModule;
	}
	
	/**
	 * @return instance of root module, defined by sysproperty "test.rootModule"
	 */
	private static Module initRootModule() {
		try {
			Class<?> moduleClass =  Class.forName(System.getProperty("test.rootModule"));
			return (Module)moduleClass.getConstructor().newInstance();
		} catch (Throwable e) {
			throw new RuntimeException("Unable to initialize Guice module", e);
		}
	}
}
````


**System properties:**
````
test.rootModule = com.my.super.org.ModuleForEnvironment1
````

**Base test class:**
````java
@Listeners({MyInjectableListenerFactory.class, MySuperListener.class})
@Guice(moduleFactory = MyModuleFactory.class)
public class BaseTest {
  .....
}

**Comment by avarabyeu** [comment]
Another option: https://gist.github.com/avarabyeu/d7651146d87aaabafa53

**Comment by 573** [comment]
@avarabyeu, I would like to try your approach. How would that exactly be used? `@Listeners({ListenerWithInject.class})` in the test class and `@UseCachedInjector` in the `ListenerWithInject` class?

**Comment by 573** [comment]
Still interested in [the option](https://github.com/cbeust/testng/issues/279#issuecomment-205449040) @avarabyeu provides was able to come up with a though a bit hackish option from my perspective:

I assume that you know how to setup guice modules the default way, it is described sufficiently elsewhere (what you bind is not the scope of #279 so adapt this option to whatever you work on, what I did serving as an example here is bind EventBus and some custom handler implementation I did provide/use to get specific informations generated by the tests run).

In the test class I have a listener:

```
@Listeners({CustomListener.class})
public class DemoITCase {
    @Inject
    EventBus eventBus;
    
    @Inject
    CustomHandler customHandler;
    
    // now this is more or less the key to it
   @BeforeTest
    public  void adaptInjectionForListener(ITestContext iTestContext) {
        iTestContext.setAttribute(customHandler.getClass().getSimpleName(), customHandler);
        eventBus.register(customHandler);
    }
    
    // etc. your tests
}
```

For completeness the listener which implements `import org.testng.IHookable;`:

```
public class VerifyContinueOnErrorResults implements IHookable {
    
    @Override
    public void run(IHookCallBack callBack, ITestResult testResult) {
        // parts irrelevant for this demo left out
        CustomHandler customHandler = (CustomHandler) testResult.getTestContext().getAttribute(CustomHandler .class.getSimpleName());
    }
}
```

@cbeust [got](https://groups.google.com/forum/#!msg/testng-users/gN1JBfG7zAQ/dj3S3nhDNpwJ) \*me this idea.

**Comment by howardzeng** [comment]
I dont unterstand about Guice dependency injection,could you give me some easy sample!

thanks very much

**Comment by juherr** [comment]
@howardzeng 

http://testng.org/doc/documentation-main.html#guice-dependency-injection
http://biggerwrench.blogspot.fr/2014/02/testng-using-guice-for-dependency.html

**Comment by ppadialitrs** [comment]
+1 any update on this?

**Comment by alinpa** [comment]
+1 any news on this?

**Comment by polaire801** [comment]
fwiw -- +1 for me as well.

**Comment by paul-hammant** [comment]
Thanks for that snippet - I'm happy to use hacks in test code :)

**Comment by krmahadevan** [comment]
@juherr - Now that TestNG basically uses only the first created instance of a listener and discards away the rest, do you think that we can give a shot at this based on the comments provided by @cbeust in https://github.com/cbeust/testng/issues/279#issuecomment-128774805

**Comment by juherr** [comment]
@krmahadevan Feel free to try but the main problem is injection will only work with the parent injector without the module provided by `@Guice` in test classes.
To be honest, I don't know if many people configure the parent injector but it could be a good start ;)

**Comment by symonk** [comment]
Any updates? it is so annoying to use a DI framework and then calling new(); in listeners etc

**Comment by polaire801** [comment]
I've experimented a little bit, and this is very hack-ish, so YMMV. 

Some of the listeners work at the suite level, as opposed to the test or class level, so you'll want to keep the test or suite's lifecycle in mind.

The modules are bound by the injector, so you can retrieve them from the parent injector. If you know which modules are used, you can specify the module's class name in the binding reference.

```
public class CustomReporter implements IReporter {

    @Override
    public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites,
        String outputDirectory) {

        Injector parentInjector = suites.get(0).getParentInjector();

        MyTestModule myModule = parentInjector.getInstance(MyTestModule.class);

        Injector childInjector = parentInjector.createChildInjector(myModule);

         SomeObjectFromModule someObject = childInjector.getInstance(SomeObjectFromModule.class);


        // do stuff

    }
}

```

```
@Listeners(CustomReporter.class)
@Guice(modules = MyTestModule.class)
public class SomeTest { }

```
Things to consider:

I would avoid using `@Inject` within a listener because of the potential changes throughout the test's lifecycle. Plus, calling `injectMembers` is slow. It would be faster to call a method directly from the module's instance, or if your binding is dynamic / just-in-time, create the child injector and then retrieve the instance (like in the example above).

To make it more dynamic, you could experiment with fetching the module names from the test class's Guice annotation via test-or-suite context. You could then retrieve the modules' instances from the parent injector and then pass them into the child injector's instance.

**Comment by symonk** [comment]
this seems to work for me:

```
@Slf4j
public class TestExecutionListener implements ITestListener {

  @Inject private SyleniumProperties properties;
  @Inject private ReportInteractable reportHelper;
  @Inject private Communicator communicator;
```

```
  @Override
  public void onStart(final ITestContext context) {
    final Injector parentInjector = context.getSuite().getParentInjector();
    final List<Module> guiceModules = context.getGuiceModules(GuiceModule.class);
    parentInjector.createChildInjector(guiceModules).injectMembers(this);
```

I don't know how it will hold up when I run multiple suites and things, but it seems ok so far, not overly keen on  the field injection but can't think of another way.

**Comment by polaire801** [comment]
@Sy That module method is sweet! I forget I'm maintaining a suite that uses
an old version. Nice!

**Comment by paul-hammant** [comment]
I myself have been looking at [Cuppa](https://github.com/cuppa-framework/cuppa) for test-running and do **all** my 'scoped container' needs for unit/service/functional hierarchies in a single execution (with hard gates). I'm also able to do everything I need without an actual DI container involved ([epiphany blog post](https://paulhammant.com/2018/04/10/it-turns-out-we-didnt-need-dependency-injection-containers-after-all/)). The trouble is that project is not very active even though I wish it was.

**Comment by juherr** [comment]
Could be great for 7.0

**Comment by Mjl33** [comment]
Would love this to be in the next release

**Comment by krmahadevan** [comment]
This issue now stands fixed and the fix will be available in the upcoming `7.5.0` release.
Now you can inject dependencies into listeners and reporters which were wired in via

1. `@Listeners` annotation
2. `<listeners>` tag

Note:
Listeners wired in via `ServiceLoader` wont be dependency injected because their instantiation is being managed by the JVM itself into which we dont have visibility.
